sdks/python/apache_beam/runners/portability/abstract_job_service.py

```diff
@@ -115,13 +127,43 @@ def __init__(self, job_id, job_name, pipeline, options):
     self._job_name = job_name
     self._pipeline_proto = pipeline
     self._pipeline_options = options
+    self._state_history = [(beam_job_api_pb2.JobState.STOPPED, time.time())]
 
   def _to_implement(self):
     raise NotImplementedError(self)
 
   prepare = run = cancel = _to_implement
   artifact_staging_endpoint = _to_implement
-  get_state = get_state_stream = get_message_stream = _to_implement
+  get_state_stream = get_message_stream = _to_implement
+
+  @property
+  def state(self):
+    """Get the latest state enum."""
+    return self.get_state()[0]
+
+  def get_state(self):
+    """Get a tuple of the latest state and its timestamp."""
+    # this is safe: initial state is set in __init__
+    return self._state_history[-1]
+
+  def set_state(self, new_state):
+    """Set the latest state as an int enum and update the state history.
+
+    :param new_state: int
+      latest state enum
+    :return: float or None
+      the new timestamp if the state has not changed, else None
+    """
+    if new_state != self._state_history[-1][0]:
+      timestamp = time.time()
```

Comment:
Consider using Beam's timestamp object instead of the native time object in Python since it already supports to/from proto: https://github.com/apache/beam/blob/b446304f75078ca9c97437e685409c31ceab7503/sdks/python/apache_beam/utils/timestamp.py#L145

This will prevent the proliferation of multiple time like objects.

Reply:
yeah, good suggestion.